Foundation/Operation.swift

```diff
@@ -234,81 +234,163 @@ extension OperationQueue {
+    private(set) weak var tail: _IndexedOperationLinkedList.Node? = nil
+    private(set) var count: Int = 0
+    private var nodeForOperation: [Operation: _IndexedOperationLinkedList.Node] = [:]
+    
+    func insert(_ operation: Operation) {
+        let node = _IndexedOperationLinkedList.Node(operation: operation)
+        
+        if root == nil {
+            root = node
+            tail = node
+        } else {
+            node.previous = tail
+            tail?.next = node
+            tail = node
+        }
+        nodeForOperation[node.operation] = node
+        count += 1
+    }
+    
+    func remove(_ operation: Operation) {
+        guard let node = nodeForOperation.removeValue(forKey: operation) else {
+            return
+        }
+        
+        guard let unwrappedRoot = root, let unwrappedTail = tail else {
+            // There cannot be a case where `nodeForOperation` contains a node and either `root` or `tail` is nil
+            // When root and tail are `nil`, `nodeForOperation` dictionary must be empty.
+            fatalError()
+        }
+        
+        if node === unwrappedRoot {
+            let next = node.next
+            next?.previous = nil
+            root = next
+            count -= 1
+            return
+        }
+        
+        if node === unwrappedTail {
+            tail = node.previous
+            tail?.next = nil
+            count -= 1
+            return
+        }
+        
+        // Middle Node
+        let previous = node.previous
+        let next = node.next
+        
+        previous?.next = next
+        next?.previous = previous
+        
+        node.next = nil
+        node.previous = nil
+        count -= 1
+    }
+    
+    func removeFirst() -> Operation? {
+        guard let returnNode = root else {
+            return nil
+        }
+        
+        remove(returnNode.operation)
+        return returnNode.operation
+    }
+    
+    func map<T>(_ transform: (Operation) throws -> T) rethrows -> [T] {
+        var result: [T] = []
+        var current = root
+        while let node = current {
+            result.append(try transform(node.operation))
+            current = current?.next
+        }
+        return result
+    }
+    
+}
+
+
 internal struct _OperationList {
-    var veryLow = [Operation]()
-    var low = [Operation]()
-    var normal = [Operation]()
-    var high = [Operation]()
-    var veryHigh = [Operation]()
-    var all = [Operation]()
+    var veryLow = _IndexedOperationLinkedList()
+    var low = _IndexedOperationLinkedList()
+    var normal = _IndexedOperationLinkedList()
+    var high = _IndexedOperationLinkedList()
+    var veryHigh = _IndexedOperationLinkedList()
+    var all = _IndexedOperationLinkedList()
+    
+    var count: Int {
+        return all.count
+    }
     
     mutating func insert(_ operation: Operation) {
-        all.append(operation)
+        all.insert(operation)
```

Comment:
What are the required threading guarantees here?

Reply:
Im not sure what's the requirement around this logic. Looking at the implementation of OperationQueue, I see all the the usage of `_operations` is wrapped by `OperationQueue.lock`

Reply:
So it is thread safe.

Reply:
@parkera Please let me know if my comments address you concern.